### DISCUSSION TITLE
HTTPX is dropping AnyIO because of performance issues, what can we learn?

### DISCUSSION THREAD

**Discussion by jruere** [agronholm/anyio#760] [General]
It's very unfortunate that HTTPX is replacing AnyIO for a native asyncio backend. They already had a native Trio backend.

Here are the discussions:
- Beginning of performance issue: https://github.com/encode/httpx/issues/3215
- New asyncio backend: https://github.com/encode/httpcore/pull/930
- Use asyncio synchronization primitives: https://github.com/encode/httpcore/pull/922

Unfortunately, improving AnyIO was not considered, AFAICS.
On the other hand, already having an abstraction for backends makes AnyIO somewhat redundant.

There's a discussion about lowering the overall cost of the AnyIO (https://github.com/agronholm/anyio/discussions/543). I don't know how much the approach discussed there would help, but I assume it would mostly help with the AnyIO backend as a whole.

What about the synchronization primitives? Where is the extra cost coming from? I doubt there's per-call backend dispatch (as there is with `anyio.sleep()`). Are the statistics the problem? Is the code just slower? Could we copy the code from asyncio for a cheap speed-up? Could we extend the asyncio primitives to get most of the benefit without adding more code?

**Comment by smurfix** [comment]
I strongly suspect that the reason for much of the delay is that anyio does a lot of work to impart Trio semantics on asyncio, which isn't built for that, and the additional work (workarounds …) don't exactly improve performance.

So httpx now has an asyncio backend and a trio backend, right? *Shrug* I don't see there's much we can do about this.

**Reply by jruere** [reply]
I don't think the decision HTTPX is making should be changed. They already try to be backend agnostic independently of AnyIO, for some reason.

Is the difference between synchronization primitives that great? They were not talking about CapacityLimiters.

It's hard to discuss this without something concrete. I hoped that the changes and benchmarks in the PRs would provide grounds for discussion, but reading them, they are all over the place. Instead of just changing the primitives, it also changes a bunch of other unrelated stuff.

I guess that the benchmark would allow us to change whatever we want and compare the overhead; there's still value in that benchmark, if there's the perception of the need to make AnyIO faster.

The reason I started this discussion is that I saw no contact between the projects over this change. I thought it was a missed opportunity.
